reMarkable 2 input support

LinusCDE opened the pull request:
This PR makes the input system compatible with the rM2.

The the rM2 the event files have changed their order. There are different ways of getting the correct file. I used @raisjn's [aproach](https://gist.github.com/raisjn/01b16286dcc4461a6643f40f8f553966) by checking the capabilities of each device (see [this issue](https://github.com/Eeems/oxide/issues/48) for more).

This basicially opens every `/dev/input/event*` file once before the first device is accessed to build a HashMap correlating all `InputDevice`s event file paths.

This should also keep working when a rM3 gets released.

One downside I found with this though is, that the Multitouch device on the rM1 has some serious delay when opening it.

```
reMarkable: /home/root/ ./input # Modified to measure detection times
Duration "/dev/input/event2" opening fd: Ok(110.121µs)
Duration "/dev/input/event2" checking caps: Ok(20.957µs)
Duration "/dev/input/event2" closing fd: Ok(39.373µs)
Duration "/dev/input/event1" opening fd: Ok(35.439177ms)  # <-- Takes long!
Duration "/dev/input/event1" checking caps: Ok(11.541µs)
Duration "/dev/input/event1" closing fd: Ok(146.286µs)
Duration "/dev/input/event0" opening fd: Ok(234.241µs)
Duration "/dev/input/event0" checking caps: Ok(10.666µs)
Duration "/dev/input/event0" closing fd: Ok(46.206µs)
Duration total: Ok(44.654802ms) # The increased time probably due to checking the system time
"/dev/input/event2" is GPIO
"/dev/input/event1" is Multitouch
"/dev/input/event0" is Wacom
Waiting for input events...
^C
```

So this will basicially add a 30-40ms delay to anyone using ANY input device with this lib.  
One mitigation would be to keep the evdev open and either pass it onto a evcontext or drop it if a timeout runs out. This would increase the complexity of code however and I'm not sure how this should best be implemented.

I'm also not sure whether the current code should be refactored into a own struct some other cleaner approach.

@raisjn, could you test [this gzipped input example](https://github.com/canselcik/libremarkable/files/5471541/input-newdebug.gz) on the rM2 as well as I don't own one. That particular binary also includes the "benchmarking" I showed in the above example to maybe see some changes on the rM2.

**NOTE:** I'll keep this PR a draft for now since some of the values also seem to have changed and something else with the touchscreen. @raisjn will probably have specifics on this. I also would appreciate any input here as to how make this update proper.

**NOTE2:** I also notices that this PR also updates the input.rs from PR #47. So this should probably be resolved first as I'm not sure how to properly separate the PRs (I should've seen this coming :|).

LinusCDE commented:
Forgot to mark this as a **DRAFT**. Please do not merge this now as testing is not done yet, contants might need adjustments and the code probably some refactoring (any input is welcome!).

raisjn commented:
it correctly identified my touches as touches and stylus as stylus. 

```
reMarkable: ~/ ./input-newdebug
Duration "/dev/input/event2" opening fd: Ok(173µs)
Duration "/dev/input/event2" checking caps: Ok(11.875µs)
Duration "/dev/input/event2" closing fd: Ok(47.134625ms)
Duration "/dev/input/event1" opening fd: Ok(152.25µs)
Duration "/dev/input/event1" checking caps: Ok(5µs)
Duration "/dev/input/event1" closing fd: Ok(39.552875ms)
Duration "/dev/input/event0" opening fd: Ok(82.625µs)
Duration "/dev/input/event0" checking caps: Ok(2.875µs)
Duration "/dev/input/event0" closing fd: Ok(39.77575ms)
Duration total: Ok(127.452125ms)
"/dev/input/event0" is GPIO
"/dev/input/event2" is Multitouch
"/dev/input/event1" is Wacom
Waiting for input events...
MultitouchEvent { event: Press { finger: Finger { tracking_id: 1, pos: Point2 [479, 501], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 1, pos: Point2 [483, 501], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 1, pos: Point2 [488, 499], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 1, pos: Point2 [494, 495], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 1, pos: Point2 [497, 494], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 1, pos: Point2 [501, 488], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Release { finger: Finger { tracking_id: 1, pos: Point2 [501, 488], pos_updated: false, last_pressed: false, pressed: false } } }
MultitouchEvent { event: Press { finger: Finger { tracking_id: 2, pos: Point2 [65535, 492], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 2, pos: Point2 [65535, 492], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 2, pos: Point2 [65535, 492], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 2, pos: Point2 [65535, 494], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 2, pos: Point2 [65535, 495], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 2, pos: Point2 [65535, 501], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 2, pos: Point2 [65535, 505], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 2, pos: Point2 [65535, 510], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 2, pos: Point2 [65535, 516], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 2, pos: Point2 [65535, 521], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 2, pos: Point2 [65535, 528], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 2, pos: Point2 [65535, 536], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 2, pos: Point2 [65535, 543], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Move { finger: Finger { tracking_id: 2, pos: Point2 [65535, 550], pos_updated: true, last_pressed: true, pressed: true } } }
MultitouchEvent { event: Release { finger: Finger { tracking_id: 2, pos: Point2 [65535, 550], pos_updated: false, last_pressed: false, pressed: false } } }
WacomEvent { event: InstrumentChange { pen: Touch, state: true } }
WacomEvent { event: InstrumentChange { pen: ToolPen, state: true } }
WacomEvent { event: Hover { position: Point2 [740.70483, 1103.0042], distance: 0, tilt: Vector2 [63136, 65336] } }
WacomEvent { event: Hover { position: Point2 [740.1691, 1102.1113], distance: 0, tilt: Vector2 [63136, 65336] } }
WacomEvent { event: Hover { position: Point2 [739.3656, 1101.0399], distance: 0, tilt: Vector2 [63136, 65336] } }
WacomEvent { event: Hover { position: Point2 [738.7406, 1100.5043], distance: 0, tilt: Vector2 [63136, 65336] } }
WacomEvent { event: Hover { position: Point2 [738.0263, 1100.1471], distance: 0, tilt: Vector2 [63136, 65336] } }
WacomEvent { event: Hover { position: Point2 [737.2228, 1099.8793], distance: 0, tilt: Vector2 [63136, 65336] } }
WacomEvent { event: Hover { position: Point2 [736.5085, 1099.7007], distance: 0, tilt: Vector2 [62936, 65336] } }
WacomEvent { event: Hover { position: Point2 [735.7049, 1099.6115], distance: 0, tilt: Vector2 [62936, 65336] } }
WacomEvent { event: Hover { position: Point2 [734.7228, 1099.6115], distance: 0, tilt: Vector2 [62936, 65336] } }
WacomEvent { event: Hover { position: Point2 [733.74066, 1099.6115], distance: 0, tilt: Vector2 [62936, 65336] } }
WacomEvent { event: Hover { position: Point2 [732.9371, 1099.6115], distance: 0, tilt: Vector2 [62936, 65336] } }
WacomEvent { event: Hover { position: Point2 [732.13354, 1099.8793], distance: 0, tilt: Vector2 [62936, 65336] } }
WacomEvent { event: Hover { position: Point2 [731.24066, 1100.1471], distance: 0, tilt: Vector2 [62936, 65336] } }
WacomEvent { event: Hover { position: Point2 [730.34784, 1100.4149], distance: 0, tilt: Vector2 [62836, 65336] } }
WacomEvent { event: Hover { position: Point2 [729.455, 1100.7721], distance: 0, tilt: Vector2 [62836, 65336] } }
WacomEvent { event: Hover { position: Point2 [728.7407, 1101.3971], distance: 0, tilt: Vector2 [62836, 65336] } }
WacomEvent { event: Hover { position: Point2 [727.93713, 1101.9327], distance: 0, tilt: Vector2 [62836, 65336] } }
WacomEvent { event: Hover { position: Point2 [727.1336, 1102.4685], distance: 0, tilt: Vector2 [62836, 65336] } }
WacomEvent { event: Hover { position: Point2 [726.33, 1103.0935], distance: 0, tilt: Vector2 [62836, 65336] } }
WacomEvent { event: Hover { position: Point2 [725.6157, 1103.7184], distance: 0, tilt: Vector2 [62836, 65336] } }
WacomEvent { event: Hover { position: Point2 [724.9015, 1104.3434], distance: 0, tilt: Vector2 [62836, 65336] } }
WacomEvent { event: Hover { position: Point2 [724.2765, 1105.0577], distance: 0, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [723.6515, 1105.772], distance: 0, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [723.1158, 1106.5754], distance: 0, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [722.5801, 1107.379], distance: 0, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [722.13367, 1108.1826], distance: 0, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [721.7765, 1108.9861], distance: 0, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [721.3301, 1109.879], distance: 0, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [720.88367, 1110.8611], distance: 0, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [720.43726, 1111.6647], distance: 0, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [720.08014, 1112.5575], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [719.63367, 1113.5396], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [719.18726, 1114.5217], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [718.65155, 1115.5038], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [718.20514, 1116.5752], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [717.66943, 1117.5574], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [717.1337, 1118.6287], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [716.5087, 1119.7001], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [715.8837, 1120.7715], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [715.16943, 1121.8429], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [714.4552, 1123.0035], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [713.5623, 1124.075], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [712.7588, 1125.2357], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [712.0445, 1126.3964], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [711.2409, 1127.557], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [710.3481, 1128.7178], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [709.45526, 1129.5212], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [708.38385, 1130.2355], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [707.4017, 1131.2177], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [706.41956, 1132.3783], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [705.43744, 1133.6283], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [704.4553, 1134.789], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [703.4732, 1135.7711], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [702.491, 1136.9318], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [701.5089, 1138.0924], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [700.3482, 1139.1638], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [699.3661, 1140.146], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [698.5625, 1141.2174], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [697.84827, 1142.1995], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [697.22327, 1143.1816], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [696.77686, 1144.2529], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [696.3304, 1145.2351], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [696.06256, 1146.1279], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [695.884, 1147.0208], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [695.7947, 1147.9136], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [695.70544, 1148.8064], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [695.884, 1149.61], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [696.15186, 1150.5028], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [696.4197, 1151.3956], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [696.9554, 1152.2885], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [697.759, 1153.092], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [698.6518, 1153.8956], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [699.9018, 1154.6099], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [701.3303, 1155.4133], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [702.8482, 1156.3955], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [704.5446, 1157.2883], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [706.241, 1158.3597], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [708.11597, 1159.4312], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [710.08026, 1160.2346], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [712.1338, 1161.0382], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [714.3659, 1161.7524], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [716.77655, 1162.4668], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [719.27655, 1163.2703], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [721.8658, 1163.9846], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [724.5443, 1164.6989], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [727.2229, 1165.5917], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [729.81213, 1166.4845], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [732.49066, 1167.1987], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [735.1692, 1168.0023], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [737.7585, 1168.8059], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [740.2584, 1169.5201], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [742.75836, 1170.3237], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [745.0798, 1171.1272], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [747.2226, 1172.02], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [749.2762, 1172.9128], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [751.1511, 1173.895], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [752.84753, 1174.8771], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [754.36536, 1175.8593], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [755.97253, 1176.8413], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [757.40106, 1177.9127], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [758.65106, 1178.9841], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [759.5439, 1180.2341], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [760.3475, 1181.3948], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [760.9725, 1182.5554], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [761.4189, 1183.8054], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [761.776, 1185.1447], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [761.9546, 1186.2161], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [762.1331, 1187.1982], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [762.1331, 1188.6267], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [762.1331, 1190.2338], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [761.9546, 1191.841], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [761.8653, 1193.3588], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [761.6867, 1194.9658], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [761.3296, 1196.7515], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [760.9725, 1198.5371], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [760.526, 1200.3228], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [760.0796, 1202.1978], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [759.5439, 1204.0728], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [758.9189, 1205.9476], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [758.3832, 1207.8226], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [757.66895, 1209.9655], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [757.04395, 1212.1082], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [756.50824, 1214.3402], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [755.88324, 1216.6616], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [755.34753, 1218.8044], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [754.9011, 1220.7687], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [754.45465, 1222.7329], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [754.09753, 1224.6971], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [753.7404, 1226.572], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [753.5618, 1228.3578], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [753.29395, 1230.3219], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [753.1154, 1232.1969], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [752.9368, 1234.1611], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [752.84753, 1235.9468], distance: 0, tilt: Vector2 [62436, 65336] } }
WacomEvent { event: Hover { position: Point2 [752.9368, 1237.7324], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [753.0261, 1239.5181], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [753.1154, 1241.3038], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [753.47253, 1243.0001], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [753.82965, 1244.3394], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [754.2761, 1245.4108], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [754.8118, 1246.9286], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [755.52606, 1248.5358], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [756.32965, 1250.0536], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [757.40106, 1251.3928], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [758.56177, 1252.8213], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [759.6332, 1254.1605], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [760.7939, 1255.5891], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [761.9546, 1256.9283], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [763.1153, 1258.089], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [764.36523, 1259.339], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [765.7045, 1260.4996], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [766.9545, 1261.571], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [768.29376, 1262.5532], distance: 0, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: InstrumentChange { pen: Touch, state: false } }
WacomEvent { event: Hover { position: Point2 [768.29376, 1262.5532], distance: 16, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [770.34735, 1262.9103], distance: 16, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [770.6152, 1262.9103], distance: 19, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [771.2402, 1263.0889], distance: 23, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [771.8652, 1263.3567], distance: 18, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [772.758, 1263.8031], distance: 14, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [773.8294, 1264.4281], distance: 14, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [774.9008, 1265.0531], distance: 14, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [776.1508, 1265.7673], distance: 14, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [777.4008, 1266.3923], distance: 14, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [778.74005, 1267.1066], distance: 15, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [780.07935, 1267.6423], distance: 16, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [781.32935, 1268.178], distance: 17, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [782.49005, 1268.7137], distance: 19, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [783.65076, 1269.0708], distance: 20, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [784.72217, 1269.3387], distance: 20, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [785.615, 1269.5172], distance: 22, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [786.4186, 1269.5172], distance: 22, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [787.1328, 1269.5172], distance: 23, tilt: Vector2 [62636, 65336] } }
WacomEvent { event: Hover { position: Point2 [787.6685, 1269.3387], distance: 26, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.1149, 1269.0708], distance: 29, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.4721, 1268.7137], distance: 31, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.7399, 1268.2673], distance: 32, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.8292, 1267.8209], distance: 34, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.9185, 1267.553], distance: 36, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.9185, 1267.2852], distance: 37, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.9185, 1266.8387], distance: 38, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.9185, 1266.3923], distance: 42, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.9185, 1265.7673], distance: 44, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.8292, 1264.4281], distance: 47, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.8292, 1262.7317], distance: 52, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.7399, 1260.7675], distance: 55, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.7399, 1258.714], distance: 59, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.7399, 1256.4819], distance: 63, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.7399, 1254.1605], distance: 68, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.65063, 1251.6606], distance: 71, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.7399, 1248.9822], distance: 76, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.7399, 1246.2144], distance: 80, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.7399, 1243.1787], distance: 83, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.8292, 1240.1431], distance: 87, tilt: Vector2 [62736, 65336] } }
WacomEvent { event: Hover { position: Point2 [788.9185, 1236.929], distance: 90, tilt: Vector2 [62836, 65336] } }
WacomEvent { event: Hover { position: Point2 [789.0078, 1234.0719], distance: 90, tilt: Vector2 [62836, 65336] } }
WacomEvent { event: Hover { position: Point2 [789.1864, 1231.4827], distance: 90, tilt: Vector2 [62836, 65336] } }
WacomEvent { event: InstrumentChange { pen: ToolPen, state: false } }
WacomEvent { event: Hover { position: Point2 [789.1864, 1231.4827], distance: 90, tilt: Vector2 [62836, 65336] } }
GPIO { event: Press { button: POWER } }
GPIO { event: Unpress { button: POWER } }
```

LinusCDE commented:
> it correctly identified my touches as touches and stylus as stylus.

Thanks for testing it out! Were the screen coords okay? They should all properly assume portrait rotation.

I'll also need to adjust some constants based on [your evtest output](https://gist.github.com/raisjn/97283a7591e933880861238c35ee44e9).

LinusCDE commented:
Also interesting that closing all devices took a long time there. Really weird and the detection time is pretty high there as well.

LinusCDE commented:
I'm thinking about adding a device and model struct instead of using evdev capabilities as this is much faster (on the rM1 only 1.5ms instead of 35-45ms and ~127ms on rM2).

For now I added a second branch ([rm2-input-aproach2](https://github.com/LinusCDE/libremarkable/tree/feature/rm2-input-aproach2)) with the latest commit adding that. (I only has the detection for now).

I'm not sure what approach to take. Should go the route of hardcoding even file paths depending on the detected model for some speed or stick with the more correct and probably future proof route in this pr? @canselcik @fenollp

raisjn commented:
The way I'm using id_by_capabilities does not require any extra open/close calls as the function takes a file descriptor and returns its type of input. The hashmap is likely unnecessary, as you can identify the file descriptor on demand.

LinusCDE commented:
> The way I'm using id_by_capabilities does not require any extra open/close calls as the function takes a file descriptor and returns its type of input. The hashmap is likely unnecessary, as you can identify the file descriptor on demand.

That's correct. But I need to at least open every event file the moment someone want's one device. Sure I can reuse it then, but this also increases the complexity of the library as I should probably time a point to close the fd if it's not required early on.

So when one programmer asks the library to e.g. get the events of the GPIO-Buttons, the library would introduce 35ms of additional delay because it also opened the event file for multitouch to figure out which is which.

fenollp commented:
> The rationale on checking just the number of /sys/devices/soc0/machine (code here) is that while the file seems to contain reMarkable 1.0 and reMarkable 2.0 respectivly, Eeems also had a reMarkable Prototype 1 or similar for his rM1. So checking the number and ensuring it isn't a digit seems most appropriate to me.

I'm a proponent of reading the first like 64 bytes of this file and explicitly checking each of these strings in whole, adding the weird/new ones when they come up, returning a usable error on no match and caching the result.

raisjn commented:
the enumerated reasoning so far is:

1. perf consideration
2. correctness and future proofing
3. difficulty of implementation

the func i supplied hits 1 and 2 better than hard coding byte signatures. i don't know how current code is laid out, but i am surprised it's difficult to plug in to the current code base. 

personally, i do not buy the supplied reasoning (when someone asks for button fd, need to open all fds to find it), as you can write the code not to be that way. it seems like the major reason to go with hard coding is ease is implementation

raisjn commented on src/framebuffer/common.rs:
```diff
@@ -14,11 +15,17 @@ pub type NativeWidthType = u32;
+    /// Will be 767 rM1 and 1403 on the rM2
+    pub static ref MTWIDTH: u16 = SCAN.mt_width;
+    /// Will be 1023 the rM1 and 1871 on the rM2
+    pub static ref MTHEIGHT: u16 = SCAN.mt_height;
 
-pub const WACOMWIDTH: u16 = 15725;
-pub const WACOMHEIGHT: u16 = 20967;
+    /// Will be 15725 on both the rM1 and rM2
+    pub static ref WACOMWIDTH: u16 = SCAN.wacom_width;
+    /// Will be 20967 on the rM1 and 20966 on the rM2
+    pub static ref WACOMHEIGHT: u16 = SCAN.wacom_height;
```
awesome. if this works, i'm going to try to do the same :D

raisjn commented on src/input/gpio.rs:
```diff
@@ -44,42 +44,36 @@ pub fn decode(ev: &input_event, outer_state: &InputDeviceState) -> Option<InputE
                 _ => return None,
             };
 
-            // Edge trigger -- debouncing
-            let new_state = ev.value != 0;
-            if new_state == before_state {
-                return None;
-            }
```
why deleted?

LinusCDE replied:
That is part of this issue and explained there: https://github.com/canselcik/libremarkable/pull/47

I didn't separate that, the new input example and this properly which is why I made that issue kinda a given for this one instead of dealing with tons of git merge/rebase conflicts.

raisjn reviewed (commented):
plz let me know if there's anything i can run on rM2 to help test this

LinusCDE commented:
@raisjn Thanks for the review. The last commit didn't change anything in those regards.

Changes:
- Outsourced the scanning now into `scan.rs`
- Added and removed reusable devices (didn't actually save anything in the end. measurements on top seem to have been reliably distorted by getting the SystemTime).
- The constants `MTWIDTH`, `MTHEIGHT`, `WACOMWIDTH`, `WACOMHEIGHT` were changed to static refs (**may break some code!**) to be retrieved from the evdevs instead. So these should always match the correct model.
- (Some commits also have explanations)

I don't think that I need to continue pursuing the other approach for detecting the model (https://github.com/canselcik/libremarkable/pull/48#issuecomment-720773617). Maybe it will be useful for adding the Framebuffer implementation.

Here is a current input binary. It now also print the constants from above (as respective resolutions) and therefore scale input correctly to the framebuffer coordinates.

[input] on the rM 1:
```
reMarkable: /home/root/ ./input 
"/dev/input/event2" is GPIO
"/dev/input/event1" is Multitouch
"/dev/input/event0" is Wacom
Multitouch resolution: 767x1023
Wacom resolution: 15725x20967
Opened input devices in 10.205804ms
Waiting for input events...
^C
```

On the rM2 the resolutions should be `1403x1871` for multitouch and `15725x20966` for wacom [as documented here](https://github.com/LinusCDE/libremarkable/blob/296edd250c46362e3dc64c61596140913282a0de/src/framebuffer/common.rs#L18).

--- 

@fenollp @canselcik What do you think about the name [EvDevsScan](https://github.com/LinusCDE/libremarkable/blob/296edd250c46362e3dc64c61596140913282a0de/src/input/scan.rs#L17). I'm not sure on it myself. I'm tempted to use something like "EvDevsResult" (probably confuses with rusts Results) or "EvDevsOutcome". Then again naming the method that scans [new()](https://github.com/LinusCDE/libremarkable/blob/296edd250c46362e3dc64c61596140913282a0de/src/input/scan.rs#L31) also seems a bit off.

LinusCDE commented on Cargo.toml:
```diff
@@ -16,7 +16,7 @@ ioctl-gen = "0.1.1"
 mmap = "0.1.1"
 rusttype = "0.8.2"
-evdev = "0.10.1"
+evdev = { git = "https://github.com/LinusCDE/evdev.git", rev = "43fb0b33994dba667ab3f1c107b7393ac3166ffe" }
```
I also had to patch the evdev lib for now. As it wasn't able to detect the multitouch resolution.

Link to the issue is in this commit (https://github.com/canselcik/libremarkable/pull/48/commits/46a341899f6eceb4458746549d8f7be23ec8d869).

fenollp commented:
Regarding `SCAN`/`EvDevsScan`: why not just `SCANNED`/`EvDevs`?

LinusCDE commented:
> Regarding `SCAN`/`EvDevsScan`: why not just `SCANNED`/`EvDevs`?

I like the idea of `SCANNED`. I personally think that `EvDevs` is a bit misleading as it doesn't really store them after the scan. But I could go with that.

LinusCDE commented:
**Module for rotations**

Since the reMarkable 2 uses other rotations for the multitouch and potentially wacom, replaced most manual axis swaps with calculations done by a new `input::rotation` module.

**Rotations specified in new(ish) device module.**

Since evdev doesn't know if a device is rotated, I added the device detection i already mentioned in https://github.com/canselcik/libremarkable/pull/48#issuecomment-720773617 of this pr.

The new device struct also has the **rotations** configured for the various generations.

https://github.com/canselcik/libremarkable/pull/48#issuecomment-721013583:
> > The rationale on checking just the number of /sys/devices/soc0/machine (code here) is that while the file seems to contain reMarkable 1.0 and reMarkable 2.0 respectivly, Eeems also had a reMarkable Prototype 1 or similar for his rM1. So checking the number and ensuring it isn't a digit seems most appropriate to me.
> 
> I'm a proponent of reading the first like 64 bytes of this file and explicitly checking each of these strings in whole, adding the weird/new ones when they come up, returning a usable error on no match and caching the result.

Good point, I now changed this in the model detection to do as you say.

**Readded concept of reusable devices**

I also readded the concept of reusable devices, I scrapped earlier. Not sure why, but today the setup time (scanning all evdevs and then start reading them) was not just 10ms no matter what. It took 90ms, readding turned that into 45ms.

That is also what made me change the name EvDevsScan to just EvDevs as suggested by @fenollp . Now it makes total sense since there may actually be devices in there (currently up to 1 second after scan).

**Testing on the rm2**

@raisjn I would be nice if you could test it again. Here is the binary: [input.zip](https://github.com/canselcik/libremarkable/files/5497873/input.zip)  
The multitouch should detect correctly. I'm not sure whether they have rotated the wacom digitizer though (but that would now be just [a simple entry](https://github.com/LinusCDE/libremarkable/blob/f728839f71db8d6e08cd8d041f15f9c1ee2d5e82/src/device.rs#L64)).

If that works, I'll consider the features finished. Another rust dev could then review this (the code is probably not the cleanest and complexity has increased) this pr should be ready.


**TODO for after this PR is done**
Updating the `battery.rs` to work on the device. Changing the name of the power_supply based on the model should do here. Somebody else can do this then as well, if interested.

fenollp commented on src/input/scan.rs:
```diff
@@ -0,0 +1,213 @@
+            let dev = evdev::Device::open(&evdev_path)
+                .unwrap_or_else(|_| panic!("Failed to scan {:?}", &evdev_path));
+            if dev.events_supported().contains(evdev::KEY) {
+                if dev.keys_supported().contains(evdev::BTN_STYLUS as usize)
```
Could a first check of `wacom_dev.is_none() && ...` help with the timings you're experiencing?

Same with the other 2  input devices.

LinusCDE replied:
Not really. We still have 3 /dev/input/eventX files and each is opened once here, figured out what it is, and the pair of `_path` and `_dev` get assigned accordingly.

The problem is, that the new hardware mixed up the oder. See: https://github.com/Eeems/oxide/issues/48#issuecomment-690830572

So that loop is opening every device and checking the capabilities to discover what device is which InputDevice.

LinusCDE commented:
@canselcik I think this could be merged for now. The code isn't the cleanest, but it'll try to be as performant as possible.

Combined with [the shim](https://github.com/ddvk/remarkable2-framebuffer/tree/master/src/client), apps written in rust, should work well with libremarkable as of now.

canselcik commented:
Great work @LinusCDE, merged.
